.teamcity/src/main/kotlin/configurations/Gradleception.kt

```diff
@@ -20,23 +26,56 @@ class Gradleception(model: CIBuildModel, stage: Stage) : BaseGradleBuildType(sta
         javaCrash = false
     }
 
+    /*
+     To avoid unnecessary rerun, what we do here is a bit complicated:
+
+     1. Build a Gradle distribution with a fixed timestamp and hash it, but never use this distribution.
+     2. Build a Gradle distribution with this hash as a version + fixed timestamp -> dogfood-first
+     3. Build a Gradle distribution using dogfood-first with this hash as a version + fixed timestamp different from the one above -> dogfood-second
+     4. Use dogfood-second to run a Gradle build which runs tests.
+
+     */
+    val dogfoodTimestamp1 = "19800101010101+0000"
+    val dogfoodTimestamp2 = "19800202020202+0000"
     val buildScanTagForType = buildScanTag("Gradleception")
     val defaultParameters = (buildToolGradleParameters() + listOf(buildScanTagForType) + "-Porg.gradle.java.installations.auto-download=false").joinToString(separator = " ")
 
-    applyDefaults(model, this, ":distributions-full:install", notQuick = true, extraParameters = "-Pgradle_installPath=dogfood-first $buildScanTagForType", extraSteps = {
-        localGradle {
-            name = "BUILD_WITH_BUILT_GRADLE"
-            tasks = "clean :distributions-full:install"
-            gradleHome = "%teamcity.build.checkoutDir%/dogfood-first"
-            gradleParams = "-Pgradle_installPath=dogfood-second -PignoreIncomingBuildReceipt=true $defaultParameters"
-        }
-        localGradle {
-            name = "QUICKCHECK_WITH_GRADLE_BUILT_BY_GRADLE"
-            tasks = "clean sanityCheck test"
-            gradleHome = "%teamcity.build.checkoutDir%/dogfood-second"
-            gradleParams = defaultParameters
-        }
-    })
+    applyDefaults(
+        model,
+        this,
+        ":distributions-full:install",
+        notQuick = true,
+        extraParameters = "-Pgradle_installPath=dogfood-first-for-hash -PignoreIncomingBuildReceipt=true -PpromotionCommitId=HEAD -PbuildTimestamp=$dogfoodTimestamp1 $buildScanTagForType",
```

Comment:
Ah, so `BUILD_COMMIT_ID` is not something Teamcity sets, but that is only set by some of our configurations?

Reply:
We read commit id in the order of `BUILD_COMMIT_ID`/`-PpromotionCommitId`/`BUILD_VCS_NUMBER`: https://github.com/gradle/gradle/blob/16006ce8a5b3765e577a78785fb461ba7c1f2946/build-logic/module-identity/src/main/kotlin/gradlebuild.module-identity.gradle.kts#L85